docs/usermanual-clusters.xml

```diff
@@ -5,306 +5,509 @@
+	  the Grapheme Cluster Boundary specification in <ulink
+	  url="https://www.unicode.org/reports/tr29/#Regex_Definitions">Unicode
+	  Technical Report 29</ulink>.
+	</para>
+	<para>
+	  Client programs can specify level 0 behavior for a buffer by
+	  setting its <literal>cluster_level</literal> to
+	  <literal>HB_BUFFER_CLUSTER_LEVEL_MONOTONE_GRAPHEMES</literal>. 
+	</para>
+      </listitem>
+      <listitem>
+	<para>
+	  <emphasis>Level 1</emphasis> tweaks the old behavior
+	  slightly to produce better results. Therefore, level 1
+	  clustering is recommended for code that is not required to
+	  implement backward compatibility with the old HarfBuzz.
+	</para>
+	<para>
+	  Level 1 differs from level 0 by not merging the 
+	  clusters of marks and other modifier code points with the
+	  preceding "base" code point's cluster. By preserving the
+	  cluster values of these marks and modifier code points,
+	  script shaping can perform additional operations that might
+	  lead to improved results (for example, reordering a sequence
+	  of marks).
+	</para>
+	<para>
+	  Client programs can specify level 1 behavior for a buffer by
+	  setting its <literal>cluster_level</literal> to
+	  <literal>HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS</literal>. 
+	</para>
+      </listitem>
+      <listitem>
+	<para>
+	  <emphasis>Level 2</emphasis> differs significantly in how it
+	  treats cluster values. In level 2, HarfBuzz never merges
+	  clusters.
+	</para>
+	<para>
+	  This difference can be seen most clearly when HarfBuzz processes
+	  ligature substitutions and glyph decompositions. In level 0 
+	  and level 1, ligatures and glyph decomposition both involve
+	  merging clusters; in level 2, neither of these operations
+	  triggers a merge.
+	</para>
+	<para>
+	  Client programs can specify level 2 behavior for a buffer by
+	  setting its <literal>cluster_level</literal> to
+	  <literal>HB_BUFFER_CLUSTER_LEVEL_CHARACTERS</literal>. 
+	</para>
+      </listitem>
+    </itemizedlist>
+    <para>
+      It is not <emphasis>required</emphasis> that the cluster values
+      in a buffer be monotonically increasing. However, if the initial
+      cluster values in a buffer are monotonic and the buffer is
+      configured to use clustering level 0 or 1, then HarfBuzz
+      guarantees that the final cluster values in the shaped buffer
+      will also be monotonic. No such guarantee is made for cluster
+      level 2.
+    </para>
+    <para>
+      In levels 0 and 1, HarfBuzz implements the following conceptual model for
+      cluster values:
+    </para>
+    <itemizedlist spacing="compact">
+      <listitem>
+	<para>
+          The sequence of cluster values will always remain monotonic.
+	</para>
+      </listitem>
+      <listitem>
+	<para>
+          Each cluster value represents a single cluster.
+	</para>
+      </listitem>
+      <listitem>
+	<para>
+          Each cluster contains one or more glyphs and one or more
+          characters.
+	</para>
```

Comment:
Good question. @behdad — does 'monotonic' == 'monotonically increasing'?

Reply:
I guess it must, otherwise you'd have to detect the difference....

Reply:
Nope.  Either non-increasing or non-decreasing.  Also, if input clusters are monotonic one way or another, and direction is backward (RTL or BTT), then output clusters are monotonic in the inverse direction.

Reply:
Wait; are you saying that if the input is mono-decreasing in a RTL buffer, then the output is mono-increasing?

Reply:
> Wait; are you saying that if the input is mono-decreasing in a RTL buffer, then the output is mono-increasing?

Obviously!  RTL reverses the buffer.

Reply:
I don't know that that's obvious to someone new, though.... It's easy to get the meaning of 'inverse' confused when we're already referring to something as 'backward'.

Reply:
> Also it should be noted that both level 0 and 1 maintain the invariant that cluster values are always in ascending order (from left to right or from right to left based on buffer direction), so clients can rely on that when processing the cluster values. (May be that what monotonic mean, but I wouldn’t know from just reading that).

This is only true / guaranteed, if the input clusters were monotonically non-decreasing.

Reply:
OK; I would prefer to merge this as it stands after this latest commit (b861302), to get the other changes out. 

I will need to add a more explicit discussion of monotony & backwardness earlier in the doc. Better to do that in one place rather than allude to it in several spots; I think that will get people up to speed quicker.